pype/lib.py

```diff
@@ -708,3 +709,607 @@ def __init__(self):
+        if not variant_loaders:
+            log.warning((
+                "Workfile variant has missing loaders configuration: {0}"
+            ).format(json.dumps(variant, indent=4)))
+            continue
+
+        found = False
+        for loader_name in variant_loaders:
+            if loader_name in loaders_by_name:
+                valid_variants.append(variant)
+                found = True
+                break
+
+        if not found:
+            log.warning(
+                "Loaders in Workfile variant are not available: {0}".format(
+                    json.dumps(variant, indent=4)
+                )
+            )
+
+    if not valid_variants:
+        log.warning("There are not valid Workfile variants. Skipping process.")
+        return
+
+    log.debug("Valid Workfile variants: {}".format(valid_variants))
+
+    subsets = []
+    version_by_subset_id = {}
+    repres_by_version_id = {}
+    for subset_id, in_data in asset_entity_data["subsets"].items():
+        subsets.append(in_data["subset_entity"])
+        version_data = in_data["version"]
+        version_entity = version_data["version_entity"]
+        version_by_subset_id[subset_id] = version_entity
+        repres_by_version_id[version_entity["_id"]] = version_data["repres"]
+
+    if not subsets:
+        log.warning("There are not subsets for asset {0}".format(
+            asset_entity["name"]
+        ))
+        return
+
+    subsets_by_family = collections.defaultdict(list)
+    for subset in subsets:
+        family = subset["data"].get("family")
+        if not family:
+            families = subset["data"].get("families")
+            if not families:
+                continue
+            family = families[0]
+
+        subsets_by_family[family].append(subset)
+
+    valid_subsets_by_id = {}
+    variants_per_subset_id = {}
+    for family, subsets in subsets_by_family.items():
+        family_low = family.lower()
+        for variant in valid_variants:
+            # Family filtering
+            variant_families = variant.get("families") or []
+            if not variant_families:
+                continue
+
+            variant_families_low = [fam.lower() for fam in variant_families]
+            if family_low not in variant_families_low:
+                continue
+
+            # Regex filtering (optional)
+            variant_regexes = variant.get("subset_name_filters")
+            for subset in subsets:
+                if variant_regexes:
+                    valid = False
+                    for pattern in variant_regexes:
+                        if re.match(pattern, subset["name"]):
+                            valid = True
+                            break
+
+                    if not valid:
+                        continue
+
+                subset_id = subset["_id"]
+                valid_subsets_by_id[subset_id] = subset
+                variants_per_subset_id[subset_id] = variant
+
+            # break variants loop if got here
```

Comment:
This comment could be improved by describing why it's breaking there as opposed to explaining that it's just breaking the loop. Now it seems to explain what the `break` command does. Should it maybe instead explain e.g. `# break variants loop on finding the first matching variant`?